base/compiler/loops.jl

```diff
@@ -0,0 +1,350 @@
+                          Expr(:call, GlobalRef(Main.Base, :getindex), SSAValue(indicesid), i),
+                          tt.parameters[i],
+                          nextid)
+    end
+    # Build the loops
+    startloc, stoploc = src.codelocs[id], src.codelocs[bodystop+1]
+    loopbody = PrePostBlock(StmtBlock(startloc),
+                            body,
+                            StmtBlock(stoploc))
+    iteritem_ids  = Array{SSAValue}(undef, N)
+    exitid = loopexit
+    # From the standpoint of marking entrances and exits, it's best to
+    # generate the loops from outer-to-inner. But nesting should go
+    # the other direction, so we temporarily store a list of
+    # independent loops.
+    loops = Vector{PrePostBlock}(undef, N)
+    for i = N:-1:1
+        # Add the `iterate(iter)` call
+        pre, enterid, item, itemtype, state, statetype, nextid =
+            iterate_block!(src, iteratorids[i],
+                           tt.parameters[i],
+                           length(src.slotnames)+1,
+                           Int(startloc), -1, Int(exitid), Int(nextid))
+        iteritem_ids[i]  = item
+        # Add the `iterate(iter, state)` call
+        post, _, _, _, _, _, nextid =
+            iterate_block!(src, iteratorids[i],
+                           tt.parameters[i],
+                           length(src.slotnames),
+                           Int(stoploc), Int(enterid), Int(exitid),
+                           Int(nextid), state, statetype)
+        loops[i] = PrePostBlock(pre, nothing, post)
+        exitid = post.ssavalueids[1]
+    end
+    # In the loop body, create the `CartesianIndex` that would have
+    # resulted from iterating the CartesianIndices
+    tupleid = SSAValue(nextid)
+    nextid = addstmt!(loopbody.pre,
+                      Expr(:call, GlobalRef(Core, :tuple), iteritem_ids...),
+                      NTuple{N,Int},
+                      nextid)
+    nextid = addstmt!(loopbody.pre,
+                      Expr(:(=), itervar, Expr(:new, Main.Base.CartesianIndex{N}, tupleid)),
+                      Main.Base.CartesianIndex{N},
+                      nextid)
+    # Nest the loops
+    outerloop = loop = loops[N]
+    for i = N-1:-1:1
+        loop.body = loops[i]
+        loop = loops[i]
+    end
+    loop.body = loopbody
+    # Concatenate back to a linear representation
+    whole = StmtBlock(wholepre.codelocdefault)
+    append!(whole, wholepre)
+    append!(whole, destruct)
+    append!(whole, outerloop)
+    append!(whole, wholepost)
+    # Fix up the numbering
+    changemap = fill(typemin(Int), nextid)
+    for i = 1:length(whole.ssavalueids)
+        id = whole.ssavalueids[i]
+        changemap[id] = i - id
+    end
+    remap_ir_elements!(whole.stmts, changemap)
+    src.code = whole.stmts
+    src.codelocs = whole.codelocs
+    src.ssavaluetypes = whole.ssavaluetypes
+    return src
+    # newsrc = CodeInfo(whole.stmts, whole.codelocs, whole.ssavaluetypes,
```

Comment:
I wasn't sure whether it would be better to do in-place modification, but currently it doesn't seem possible to construct a CodeInfo from Julia. Obviously fixable via `ccall`s, but I took the lazy way out so far.